### PR TITLE
change default uid

### PR DESCRIPTION
**Pull request by huguesBouvier**
This PR address bug# 9701835.

- Start edgeAgent with default uid 13622
- Start edgeHub with default uid 13623

Tested on E2E test pipeline.

Tested on local setup manually to check backward compatibility.

arm images test:
- agent and hub running correctly on arm32 target. Tested backward compatibility scenario
- agent and hub running correctly on arm64 target. Tested backward compatibility scenario

Centos script tested manually on centos image.

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh [hunk 1]
```diff
@@ -65,6 +52,9 @@ then
     chown -fR "${TARGET_UID}" "${backuppath}"
   fi
 
+  # add user to iotedge group
+  usermod -aG iotedge "${TARGET_UID}" >/dev/null
```
Comment (darobs) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh:56]: This is running in the context of the container - do we have an iotedge group there?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh [hunk 2]
```diff
@@ -17,33 +17,20 @@
...
 # the code, so a writeable user directory is required.
 ###############################################################################
 echo "$(date --utc +"%Y-%m-%d %H:%M:%S %:z") Starting Edge Agent"
 
-TARGET_UID="${EDGEAGENTUSER_ID:-1000}"
+TARGET_UID="${EDGEAGENTUSER_ID:-13622}"
 cuid=$(id -u)
 
 if [ $cuid -eq 0 ]
 then
```
Comment (darobs) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh:33]: The adduser was placed here if a user decides to use a EDGEAGENTUSER_ID!=13622 in the module spec. Can we confirm we don't need it?

**Review by darobs — COMMENTED** [review]
Would it make sense to update [aziot-edged.mgmt.socket](https://github.com/Azure/iotedge/blob/master/edgelet/contrib/systemd/debian/aziot-edged.mgmt.socket)?

You should also update the ARM base Dockerfiles for both agent and hub. [example](https://github.com/Azure/iotedge/blob/master/edge-agent/docker/linux/arm32v7/base/Dockerfile)

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh (hunk 1 above)
Comment (arsing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh:56]: Yes, this should only be done in package preinst, not here.

Comment (huguesBouvier) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh]: Fixed

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh (hunk 2 above)
Comment (huguesBouvier) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh]: You're right, fixed.

File: edge-agent/docker/linux/arm32v7/base/Dockerfile [hunk 3]
```diff
@@ -9,7 +9,7 @@ RUN apt-get update && \
 
 # create a user to allow agent to optionally run as non-root
 ARG EDGEAGENTUSER_ID
-ENV EDGEAGENTUSER_ID ${EDGEAGENTUSER_ID:-1000}
+ENV EDGEAGENTUSER_ID ${EDGEAGENTUSER_ID:-13622}
```
Comment (darobs) [inline edge-agent/docker/linux/arm32v7/base/Dockerfile:12]: Assuming you created a new base image for master, the Dockerfile one directory up should be updated with that new image tag.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/hubStart.sh [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+###############################################################################
+# Set up EdgeHub to run as a non-root user at runtime, if allowed.
+# 
+# If this script is started as root:
+#  1. It reads the EDGEHUBUSER_ID environment variable, default UID=13623.
+#  2. If the User ID does not exist as a user, create it.
+#  3. If "StorageFolder" env variable exists, use as basepath, else use /tmp
+#     Do same for backuppath
+#  4. If basepath/edgehub exists, make sure all files are owned by EDGEHUBUSER_ID
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/hubStart.sh:11]: I don't believe EdgeHub uses backup..

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/hubStart.sh [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+#     Do same for backuppath/edgehub_backup
+#  5. Make sure /app/backup.json is writeable.
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/hubStart.sh:13]: This is not used in EdgeHub

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/hubStart.sh (hunk 4 above)
Comment (huguesBouvier) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/hubStart.sh:11]: There was this file: https://github.com/Azure/iotedge/blob/master/edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/Constants.cs#L9
So I supposed it has one.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/hubStart.sh (hunk 5 above)
Comment (huguesBouvier) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/hubStart.sh]: Forgot to remove the comment. It is removed in the script.

File: edge-agent/docker/linux/arm32v7/base/Dockerfile (hunk 3 above)
Comment (huguesBouvier) [reply edge-agent/docker/linux/arm32v7/base/Dockerfile:12]: Done!

File: edgelet/contrib/centos/aziot-edge.spec [hunk 6]
```diff
@@ -89,6 +91,11 @@ if /usr/bin/getent group systemd-journal >/dev/null; then
     %{_sbindir}/usermod -aG systemd-journal %{iotedge_user}
 fi
 
+# Create an edgeagentuser and add it to iotedge group
+if ! /usr/bin/getent passwd %{iotedge_agent_user} >/dev/null; then
+    %{_sbindir}/useradd -g %{iotedge_group} -c "edgeAgent user" -ms /bin/bash -u %{iotedge_agent_uid} %{iotedge_agent_user}
```
Comment (arsing) [inline edgelet/contrib/centos/aziot-edge.spec:96]: Login shell should be `/bin/nologin` just like the `iotedge` user. There's no need for anyone to log in as `edgeagentuser`.

File: edge-hub/docker/linux/arm64v8/Dockerfile [hunk 7]
```diff
@@ -15,8 +15,11 @@ EXPOSE 8883/tcp
...
-USER edgehubuser
 ENV OptimizeForPerformance false
 ENV MqttEventsProcessorThreadCount 1
+
+COPY Microsoft.Azure.Devices.Edge.Hub.Service/ ./
+COPY $EXE_DIR/Microsoft.Azure.Devices.Edge.Hub.Service/hubStart.sh ./
+
 CMD echo "$(date --utc +"%Y-%m-%d %H:%M:%S %:z") Starting Edge Hub" && \
-    /usr/local/bin/watchdog
+    exec /app/hubStart.sh
```
Comment (arsing) [inline edge-hub/docker/linux/arm64v8/Dockerfile:25]: Missing newline at end of line.

File: edgelet/contrib/debian/preinst [hunk 8]
```diff
@@ -12,6 +12,11 @@ add_groups()
     fi
     mkdir -p /var/lib/aziot/edged
 
+    # Create an edgeagentuser and add it to iotedge group
+    if ! getent passwd edgeagentuser >/dev/null; then
+        useradd -g iotedge -c "edgeAgent user" -ms /bin/bash -u 13622 edgeagentuser
```
Comment (arsing) [inline edgelet/contrib/debian/preinst:17]: Same here. Login shell should be `/sbin/nologin`

**Review by arsing — CHANGES_REQUESTED** [review]

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/hubStart.sh (hunk 4 above)
Comment (varunpuranik) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/hubStart.sh:11]: Hmm, yes you are right actually - I forgot about this one. It is used for the backup and restore feature...

File: edgelet/contrib/debian/preinst (hunk 8 above)
Comment (varunpuranik) [inline edgelet/contrib/debian/preinst:17]: Why is this needed here? How do we make sure the UID here is same as in the EA docker file?

File: edge-hub/docker/linux/amd64/Dockerfile [hunk 9]
```diff
@@ -12,7 +14,8 @@ RUN apk update && \
     apk add --no-cache snappy libcap	
 
 # Add an unprivileged user account for running Edge Hub	
-ARG EDGEHUBUSER_ID=1000	
+ARG EDGEHUBUSER_ID
```
Comment (varunpuranik) [inline edge-hub/docker/linux/amd64/Dockerfile:17]: This seems subtly different from this one - https://github.com/Azure/iotedge/pull/4998/files#diff-c53eb6c91226f5467c44ca4bea9845150867a5792a5a66c1d9b08d02d21ca46eR5
Just curious... Why is that?

Comment (huguesBouvier) [reply edge-hub/docker/linux/amd64/Dockerfile:17]: I changed the base image in the PR from 8 to 11. But base image had been changed on master so I had a merge conflict. Fixing it on 40 files was painful so I reverted, rebased and changed the base image again.

File: edgelet/contrib/debian/preinst (hunk 8 above)
Comment (huguesBouvier) [reply edgelet/contrib/debian/preinst:17]: Dockerfile is present for backward compatibility only. It will go away eventually.
The uuid in docker file is also harcoded.

Our user can change ithat 13622 in dockerfile but it is ok. the mgmt ownership is also re-assigned in dockerfile so it still works.

File: edge-hub/docker/linux/arm64v8/Dockerfile (hunk 7 above)
Comment (huguesBouvier) [reply edge-hub/docker/linux/arm64v8/Dockerfile:25]: fixed

File: edgelet/contrib/centos/aziot-edge.spec (hunk 6 above)
Comment (huguesBouvier) [reply edgelet/contrib/centos/aziot-edge.spec:96]: /sbin/nologin you mean? Fixed

File: edgelet/contrib/debian/preinst (hunk 8 above)
Comment (huguesBouvier) [reply edgelet/contrib/debian/preinst:17]: fixed

File: edgelet/contrib/centos/aziot-edge.spec (hunk 6 above)
Comment (arsing) [reply edgelet/contrib/centos/aziot-edge.spec]: I said `/bin/login` because that's what the `iotedge` user above uses.

Comment (huguesBouvier) [reply edgelet/contrib/centos/aziot-edge.spec]: You're right, let's be consistent.

**Review by arsing — APPROVED** [review]

**Review by darobs — APPROVED** [review]
